### ISSUE TITLE
S3 open file in append mode

### ISSUE THREAD

**Issue by fj-morales** [irods/python-irodsclient#467]
Hi, 

The definition of the open() function at:

https://github.com/irods/python-irodsclient/blob/bb03bb5bccf650838ae6468e6f90c60325a07997/irods/manager/data_object_manager.py#L299C1-L299C84

allows for opening a file in **append** mode. This is completely fine in the case of a local filesystem on disk, but S3 object store does not support append operation, which would lead to corrupted data if the open() function is used in append mode.

**Comment by trel** [comment]
Is this just a documentation exercise?   Is the python-irodsclient supposed to know whether the target resource being opened in iRODS is an s3-style object storage device?

**Comment by trel** [comment]
Maybe the iRODS s3 plugin already handles this?

https://github.com/irods/irods_resource_plugin_s3/blob/2117e1eaaff8cd29971dd410b1a03cd60745328a/s3/s3_operations.cpp#L751-L755

Have you demonstrated a failure with the python-irodsclient that we can reproduce?

**Comment by korydraughn** [comment]
How the open mode is handled is entirely on the s3 resource plugin. The PRC is not required to care what the storage is. This is true for any iRODS client.

to understand how append is handled requires looking at the resource plugin rather than the client. If there isn't documentation for that, then we need to consider what parts are important and document those.

@trel That is certainly an important part of the story.

**Comment by d-w-moore** [comment]
@fj-morales , all, please note that this isn't a true append and does not seek to file-end the way a normal POSIX 'a' mode open flag would.  What it does signal  is  to open a file for write access, without truncation.

**Comment by d-w-moore** [comment]
To expound a little further on the subject:   The meaning of ses.data_objects.open(name,'w') is an open for truncation, unlike with 'a'. Otherwise - no difference.

**Comment by korydraughn** [comment]
So it doesn't pass the append bit/flag to the server?

**Comment by d-w-moore** [comment]
Wrongly or (for S3 perhatps) rightly, it doesn't.  There's no mention of O_APPEND in the python-irodsclient. I'll just say I believe it has always been this way. 
If you wish to append to a data object with the Python client library as-is, you must seek(0,SEEK_END) and then write. open(_,'a') followed directly by write will start writing at the file beginning.

**Comment by trel** [comment]
That is a 'TODO' in the linked code...

https://github.com/irods/python-irodsclient/blob/bb03bb5bccf650838ae6468e6f90c60325a07997/irods/manager/data_object_manager.py#L318C15-L318C16

>         # TODO: Use seek_to_end

**Comment by d-w-moore** [comment]
Maybe O_APPEND couldn't be passed to a server at the time.

**Comment by d-w-moore** [comment]
Is also unclear what the todo commenter meant at the time by this:  use seek_to_end. Did he mean to pass an OAPPEND or perform a seek with SEEK_END? The latter approach seems to be what they meant literally but it constitutes a race condition because it doesn't actually guarantee an append.

**Comment by trel** [comment]
> but it constitutes a race condition 

That's interesting.  Can you show in a sequence of events what the race condition is?   I'm not sure I follow.

Is it a race condition with another, independent open call?

**Comment by d-w-moore** [comment]
On second thought , it probably is not going to cause a problem within iRODS since we lock out competing writers until we're done... in general with filesystems that's not the case however.

**Comment by trel** [comment]
Got it - okay.

**Comment by fj-morales** [comment]
> Have you demonstrated a failure with the python-irodsclient that we can reproduce?

Yes, let me arrange the code I'm using to reproduce it and paste it here.

**Comment by trel** [comment]
@fj-morales any news on this a couple years later?

**Comment by d-w-moore** [comment]
It's possible this is resolved with d8e3dc1

Will test.

**Comment by d-w-moore** [comment]
> It's possible this is resolved with [d8e3dc1](https://github.com/irods/python-irodsclient/commit/d8e3dc14e3723362bbf33f28691490781e8da96d)
> 
> Will test.

The following now seems to work as expected due to the commit mentioned above, so I guess we'd probably be ok to close this.
```
data_object_path = "..."
s3_resource_name = "..."
import irods.helpers
import irods.keywords as kw
s = irods.helpers.make_session()
with s.data_objects.open(data_object_path, 'a', **{kw.RESC_NAME_KW: s3_resource_name) as f:
    f.write(b'this text will be appended\n')
```

**Comment by korydraughn** [comment]
Does use of `a` append data to the end of the file across multiple writes? For example:

```py
with s.data_objects.open(data_object_path, 'a', **{kw.RESC_NAME_KW: s3_resource_name}) as f:
    f.write(data1)

    f.seek(0, SEEK_BEG) # Reset file write position to 0.
    f.write(data2)      # Expected to be written to the end of the S3 object.

    f.seek(0, SEEK_BEG) # Reset file write position to 0.
    f.write(data3)      # Expected to be written to the end of the S3 object.
```

**Comment by d-w-moore** [comment]
No, almost certainly not.  That is of course an un-POSIX-like behavior. Would O_APPEND sent to the server at OPEN time maybe rectify that?

**Comment by d-w-moore** [comment]
However, at least we know that after iRODS 4.2.9, you won't have interference and race conditions from other agents that would allow the catalog-stored "EOF" size to passively become out of sync with the in-memory accounting of the data object size, so at least we're deterministic in that way.

**Comment by korydraughn** [comment]
> That is of course an un-POSIX-like behavior.

What part are you referring to as being un-POSIX-like?

> Would O_APPEND sent to the server at OPEN time maybe rectify that?

Yes, that is the expectation. iRODS implements a POSIX-like interface, and because of that, people are likely to assume POSIX-y behavior.

The client is expected to pass `O_APPEND` like any other flag. We let the resource plugins interpret those flags. Here's how dstream's native transport translates flags for `open()`.
- https://github.com/irods/irods/blob/30c0349b8fe01ca70ba0e65e43b9e0e69ed18040/lib/core/include/irods/transport/default_transport.hpp#L313-L341
- https://github.com/irods/irods/blob/30c0349b8fe01ca70ba0e65e43b9e0e69ed18040/lib/core/include/irods/transport/default_transport.hpp#L357
- https://github.com/irods/irods/blob/30c0349b8fe01ca70ba0e65e43b9e0e69ed18040/lib/core/include/irods/transport/default_transport.hpp#L369

**Comment by d-w-moore** [comment]
Kory, I'll see what happens if we pass O_APPEND for  the "a" flag in open( ).


> What part are you referring to as being un-POSIX-like?

The one where the file pointer obeys a seek to  SEEK_BEG+0 instead of forcing all writes to append to the current end of file.  That is the current behavior of PRC with `dobj.open(...,"a")`.

**Comment by korydraughn** [comment]
That example/test was meant to highlight that the write operations always append to the end of the replica. That is, the seek operations do not override the append flag.

**Comment by d-w-moore** [comment]
> That example/test was meant to highlight that the write operations always append to the end of the replica. That is, the seek operations do not override the append flag.

I see.

**Comment by d-w-moore** [comment]
> Hi,
> 
> The definition of the open() function at:
> 
> https://github.com/irods/python-irodsclient/blob/bb03bb5bccf650838ae6468e6f90c60325a07997/irods/manager/data_object_manager.py#L299C1-L299C84
> 
> allows for opening a file in **append** mode. This is completely fine in the case of a local filesystem on disk, but S3 object store does not support append operation, which would lead to corrupted data if the open() function is used in append mode.
> 
> I haven't been able to identify yet how the open mode is handled in the irods command line client, but I suppose that there the issue is addressed differently.

Currently the Python iRODS Client does not apply O_APPEND in open operations, so this shouldn't be an issue!  Thanks for bringing it to our attention, you may notice by other discussions on this page the larger issue of when and how to apply the O_APPEND flag could be a germane concern!

**Comment by d-w-moore** [comment]
> That example/test was meant to highlight that the write operations always append to the end of the replica. That is, the seek operations do not override the append flag.

Interesting...

Well, re-reading now the description and considering the input you've just given, @korydraughn , I am wondering now whether the client needs to predict what replica will be opened and apply the O_APPEND flag to the operation depending on which resource type (ie, S3 or not) is likely to be involved.  I guess, if we decide to do it that way, we'd need to open another issue for that specifically.  

In any case though, I think given my last response to @fj-morales a few minutes ago - if I'm correct - we can probably close this one in any case.
